pkg/volume/glusterfs/glusterfs.go

```diff
@@ -315,3 +325,154 @@ func getVolumeSource(
 
 	return nil, false, fmt.Errorf("Spec does not reference a Gluster volume type")
 }
+
+func (plugin *glusterfsPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	if len(options.AccessModes) == 0 {
+		options.AccessModes = plugin.GetAccessModes()
+	}
+	return plugin.newProvisionerInternal(options)
+}
+
+func (plugin *glusterfsPlugin) newProvisionerInternal(options volume.VolumeOptions) (volume.Provisioner, error) {
+	return &glusterfsVolumeProvisioner{
+		glusterfsMounter: &glusterfsMounter{
+			glusterfs: &glusterfs{
+				plugin: plugin,
+			},
+		},
+		options: options,
+	}, nil
+}
+
+type glusterfsClusterConf struct {
+	glusterep          string `json:"endpoint"`
+	glusterRestvolpath string `json:"path"`
+	glusterRestUrl     string `json:"resturl"`
+	glusterRestAuth    bool   `json:"restauthenabled"`
+	glusterRestUser    string `json:"restuser"`
+	glusterRestUserKey string `json:"restuserkey"`
+}
+
+type glusterfsVolumeProvisioner struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+	options volume.VolumeOptions
+}
+
+func (plugin *glusterfsPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec)
+}
+
+func (plugin *glusterfsPlugin) newDeleterInternal(spec *volume.Spec) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Glusterfs == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.Spec.Glusterfs is nil")
+	}
+	return &glusterfsVolumeDeleter{
+		glusterfsMounter: &glusterfsMounter{
+			glusterfs: &glusterfs{
+				volName: spec.Name(),
+				plugin:  plugin,
+			},
+			path: spec.PersistentVolume.Spec.Glusterfs.Path,
+		}}, nil
+}
+
+type glusterfsVolumeDeleter struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+}
+
+func (d *glusterfsVolumeDeleter) GetPath() string {
+	name := glusterfsPluginName
+	return d.plugin.host.GetPodVolumeDir(d.glusterfsMounter.glusterfs.pod.UID, strings.EscapeQualifiedNameForDisk(name), d.glusterfsMounter.glusterfs.volName)
+}
+
+func (d *glusterfsVolumeDeleter) Delete() error {
+	var err error
+	glog.V(2).Infof("glusterfs: delete volume :%s ", d.glusterfsMounter.path)
+	volumetodel := d.glusterfsMounter.path
+	d.glusterfsClusterConf = d.plugin.clusterconf
+	newvolumetodel := dstrings.TrimPrefix(volumetodel, volprefix)
```

Comment:
This should not be done, in case that Heketi changes its volume name.  Instead when the volume is created, the Volname to ID mapping should be saved by the client.  Most likely in this client, it should be saved in etcd.